### PR TITLE
Add some instruction simplifications

### PR DESCRIPTION
**Pull request by eeckstein**
* removes an `enum` - `switch_enum` pairs
* constant fold null pointer checks against string literals

This enables to put `StaticString("some string literal").utf8Start` into a statically initialized global

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Review by kubamracek — APPROVED** [review]

File: test/IRGen/section_structs.swift [hunk 1]
```diff
@@ -38,8 +38,14 @@ struct MyStruct5 {
 }
 @_section("__TEXT,__mysection") var g_MyStruct5: MyStruct5 = MyStruct5(q: MyStruct4(a: 42, s: MyStruct1(a: 43, b: 44)), r: MyStruct4(a: 42, s: MyStruct1(a: 43, b: 44)))
 
+@_section("__TEXT,__mysection") let utf8OfStaticString = StaticString("hello").utf8Start
```
Comment (finagolfin) [inline test/IRGen/section_structs.swift:41]: @eeckstein, strangely this broke on the 32-bit Android armv7 CI but not on the aarch64 CI:
```
/home/ubuntu/jenkins/workspace/oss-swift-RA-linux-ubuntu-16.04-android/swift/test/IRGen/section_structs.swift:41:37: error: global variable must be a compile-time constant to use @_section attribute
@_section("__TEXT,__mysection") let utf8OfStaticString = StaticString("hello").utf8Start
                                    ^
```
This compiler on the Android CI is built with a full bootstrap on linux, as that CI has no prebuilt Swift compiler installed, so the swift-syntax parser is not built. Maybe that has something to do with it, but weird that it only hits for 32-bit compilation.

Comment (eeckstein) [reply test/IRGen/section_structs.swift:41]: @finagolfin do you have any error log or a link to the ci failure?

Comment (finagolfin) [reply test/IRGen/section_structs.swift:41]: Sorry, [copied that but forgot to paste it](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android/6157/consoleText).

Comment (eeckstein) [reply test/IRGen/section_structs.swift:41]: Let's disable the test for 32-bit

Comment (finagolfin) [reply test/IRGen/section_structs.swift:41]: SGTM

Comment (eeckstein) [reply test/IRGen/section_structs.swift:41]: https://github.com/apple/swift/pull/70540

**Pull request merged**